### DISCUSSION TITLE
Can't understand why component label doesn't rerender if grandparent object is updated

### DISCUSSION THREAD

**Discussion by henry-cotton** [solidjs/solid#912] [Q&A]
I'm writing a dashboard app to update some external data. The app has a screen where some buttons are output depending on whether a property is enabled/disabled. All my data is currently held in a createStore proxy called "store".

```javascript
export const SiteRule = ({ type }) => {
	return (
		<div class="rule">
			<span>
				{type} Sync: {store.domains[store.site][`${type}Rule`]["State"]}
			</span>
			<span>Last Run: </span>
			<button
				onClick={() =>
					store.domains[store.site][`${type}Rule`]["State"] ===
					"ENABLED"
						? disableRule(
								store.domains[store.site][`${type}Rule`]["Name"]
						  )
						: enableRule(
								store.domains[store.site][`${type}Rule`]["Name"]
						  )
				}>
				{store.domains[store.site][`${type}Rule`]["State"] === "ENABLED"
					? "Disable"
					: "Enable"}{" "}
				Sync
			</button>
		</div>
	);
};
```

The button makes an external Rest call, and on success gets an updated list of data, which is merged into the store. However, I couldn't get the labels to rerender when I was setting the grandparent store property equal to an updated array

```javascript
export const updateRules = async () => {
	return await restFetch("get-rules").then((result) => {
		const rules = ["order", "category", "image", "product"];
		let domains = unwrap(store.domains);
		let i = Object.keys(domains).length;
		while (i--) {
			let key = Object.keys(domains)[i];
			let l = rules.length;
			while (l--) {
				let rule = result.rules.find(
					(r) =>
						r.Name ===
						`${domains[key].siteKey.replaceAll(".", "-")}-${
							rules[l]
						}-sync`
				);
				domains[key][`${rules[l]}Rule`] =
					typeof rule !== "undefined" ? rule : false;
			}
		}
		setStore({ domains: domains });
	});
};
```

Updating store.domains as a whole didn't seem to rerender, but after some reading, I decided to try an alternate version where I updated store.domains in chunks, the properties of the array items

```javascript
export const updateRules = async () => {
	return await restFetch("get-rules").then((result) => {
		const rules = ["order", "category", "image", "product"];
		let domains = unwrap(store.domains);
		let i = Object.keys(domains).length;
		while (i--) {
			let key = Object.keys(domains)[i];
			let l = rules.length;
			while (l--) {
				let rule = result.rules.find(
					(r) =>
						r.Name ===
						`${domains[key].siteKey.replaceAll(".", "-")}-${
							rules[l]
						}-sync`
				);
				setStore("domains", key, { [`${rules[l]}Rule`]: rule });
			}
		}
	});
};
```

This then caused the labels to update.
However, I'm struggling to understand why this is the case - in both versions, I'm reading a property of the stored object, and not updating the property itself.

**Comment by Brendan-csel** [comment]
Complete guess here but at the bottom of the second code snippet, does changing...
``` js
setStore({ domains: domains })
```

...to...
``` js
setStore('domains', domains )
```

make any difference?

I'm wondering if you are swapping out the whole content of the store (which IIRC is not observable).

**Comment by otonashixav** [comment]
`unwrap` doesn't make a copy of the array, it gives you the underlying array. You're modifying it, then calling `setStore` with the modified array. However since you've already modified it, when `setStore` diffs the two arrays (which are the same array), it sees that nothing has changed between its array and the array you've passed it (since once again, they are the same array that has already been modified), so it doesn't trigger any updates. Try `let domains = [...store.domains];` instead, or better yet use `produce` e.g. `setStore("domains", produce((domains) => { ... })`.
